File: C++/1-Arrays&Hashing/Algorithms/3-HashMap-SetOperations/HashMap-SetOperations.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <set> // For std::set (Approach 5)

using namespace std;

// Approach 1: Using unordered_set for membership test
bool containsElement_unordered_set(const vector<int>& arr, int target) {
    // Create an unordered_set from the vector.  This is efficient for lookups.
    unordered_set<int> numSet(arr.begin(), arr.end());

    // Check if the target element is in the set.
    return numSet.find(target) != numSet.end();
}

// Approach 2: Using unordered_map for counting occurrences
unordered_map<int, int> countOccurrences_unordered_map(const vector<int>& arr) {
    // Create an unordered_map to store the frequency of each element.
    unordered_map<int, int> counts;

    // Iterate through the array and update the counts.
    for (int num : arr) {
        counts[num]++;
    }
    return counts;
}

// Approach 3: Find common elements between two vectors using unordered_set
vector<int> findCommonElements_unordered_set(const vector<int>& arr1, const vector<int>& arr2) {
    // Store elements of arr1 in a set for efficient lookup.
    unordered_set<int> set1(arr1.begin(), arr1.end());
    vector<int> commonElements;

    // Iterate through arr2 and check if each element is in set1.
    for (int num : arr2) {
        if (set1.find(num) != set1.end()) {
            commonElements.push_back(num);
            set1.erase(num); //prevent duplicates in result if arr2 has duplicate common elements
        }
    }
    return commonElements;
}

// Approach 4: Check for duplicates in a vector using unordered_set
bool hasDuplicates_unordered_set(const vector<int>& arr) {
    // Use a set to keep track of seen elements.
    unordered_set<int> seen;

    // Iterate through the array.
    for (int num : arr) {
        // If the element is already in the set, it's a duplicate.
        if (seen.find(num) != seen.end()) {
            return true;
        }
        // Otherwise, add the element to the set.
        seen.insert(num);
    }
    // If the loop finishes without finding a duplicate, return false.
    return false;
}

// Approach 5: Using std::set for sorted unique elements
vector<int> getUniqueSorted_set(const vector<int>& arr) {
    // Create a std::set, which automatically stores elements in sorted order and only keeps unique elements.
    set<int> uniqueSortedSet(arr.begin(), arr.end());

    // Convert the set to a vector for easier use.
    vector<int> result(uniqueSortedSet.begin(), uniqueSortedSet.end());
    return result;
}

int main() {
    // Example usage of the functions.
    vector<int> numbers = {1, 5, 2, 8, 5, 9, 2, 10};
    vector<int> numbers2 = {2, 5, 11, 15, 5};
    int target = 5;

    // Approach 1: Check if 'numbers' contains 'target'.
    cout << "Approach 1: Contains " << target << "? "
              << (containsElement_unordered_set(numbers, target) ? "Yes" : "No") << endl;

    // Approach 2: Count occurrences of each number in 'numbers'.
    cout << "Approach 2: Occurrences: " << endl;
    unordered_map<int, int> counts = countOccurrences_unordered_map(numbers);
    for (const auto& pair : counts) {
        cout << pair.first << ": " << pair.second << endl;
    }

    // Approach 3: Find common elements between 'numbers' and 'numbers2'.
    cout << "Approach 3: Common elements: ";
    vector<int> commonElements = findCommonElements_unordered_set(numbers, numbers2);
    for (int num : commonElements) {
        cout << num << " ";
    }
    cout << endl;

    // Approach 4: Check if 'numbers' has duplicates.
    cout << "Approach 4: Has duplicates? "
              << (hasDuplicates_unordered_set(numbers) ? "Yes" : "No") << endl;

     // Approach 5: Get unique sorted elements from 'numbers'.
    cout << "Approach 5: Unique sorted elements: ";
    vector<int> uniqueSorted = getUniqueSorted_set(numbers);
    for (int num : uniqueSorted) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: Java/1-Arrays&Hashing/Algorithms/3-HashMap-SetOperations/HashMap-SetOperations.java
import java.util.*;
import java.lang.*;

class HashMapSetOperations {

    public static void main(String[] args) {
        // Test data
        String[] words1 = {"apple", "banana", "apple", "orange", "banana", "kiwi"};
        String[] words2 = {"grape", "apple", "kiwi", "mango", "banana"};

        System.out.println("Original Arrays:");
        System.out.println("Array 1: " + Arrays.toString(words1));
        System.out.println("Array 2: " + Arrays.toString(words2));
        System.out.println();

        // 1. Find unique elements in an array
        System.out.println("1. Unique elements in Array 1:");
        findUniqueElements(words1);
        System.out.println();

        // 2. Find common elements between two arrays
        System.out.println("2. Common elements between Array 1 and Array 2:");
        findCommonElements(words1, words2);
        System.out.println();

        // 3. Count word frequencies in an array
        System.out.println("3. Word frequencies in Array 1:");
        countWordFrequencies(words1);
        System.out.println();

        // 4. Check if one array is a subset of another
        System.out.println("4. Is Array 2 a subset of Array 1?");
        System.out.println(isSubset(words1, words2));
        System.out.println();

        // 5. Find the first non-repeating element in an array
        System.out.println("5. First non-repeating element in Array 1:");
        findFirstNonRepeating(words1);
    }

    // 1. Find unique elements in an array using HashSet
    //    -  Time Complexity: O(n), where n is the length of the array.
    //    -  Space Complexity: O(n), in the worst case, to store all unique elements.
    public static void findUniqueElements(String[] arr) {
        Set<String> uniqueSet = new HashSet<>();
        for (String word : arr) {
            uniqueSet.add(word); // Add each word to the set.  Duplicates are automatically ignored.
        }
        System.out.println(uniqueSet); // Print the set, which contains only unique elements.
    }

    // 2. Find common elements between two arrays using HashSet
    //    -  Time Complexity: O(m + n), where m and n are the lengths of the two arrays.
    //    -  Space Complexity: O(min(m, n)), to store the smaller array in the set.
    public static void findCommonElements(String[] arr1, String[] arr2) {
        Set<String> set1 = new HashSet<>(Arrays.asList(arr1)); // Store elements of arr1 in a HashSet.
        Set<String> commonSet = new HashSet<>(); // Store the common elements
        for (String word : arr2) {
            if (set1.contains(word)) { // Check if the element from arr2 is present in set1.
                commonSet.add(word); // If present, add it to the common set.
            }
        }
        System.out.println(commonSet); // Print the common elements.
    }

    // 3. Count word frequencies in an array using HashMap
    //    -  Time Complexity: O(n), where n is the length of the array.
    //    -  Space Complexity: O(k), where k is the number of unique words.
    public static void countWordFrequencies(String[] arr) {
        Map<String, Integer> wordCountMap = new HashMap<>();
        for (String word : arr) {
            // If the word is already in the map, increment its count; otherwise, add it with a count of 1.
            wordCountMap.put(word, wordCountMap.getOrDefault(word, 0) + 1);
        }
        // Iterate through the map and print each word with its frequency.
        for (Map.Entry<String, Integer> entry : wordCountMap.entrySet()) {
            System.out.println(entry.getKey() + ": " + entry.getValue());
        }
    }

    // 4. Check if one array is a subset of another using HashSet
    //    -  Time Complexity: O(m + n), where m and n are the lengths of the two arrays.
    //    -  Space Complexity: O(m), where m is the length of the first array.
    public static boolean isSubset(String[] arr1, String[] arr2) {
        Set<String> set1 = new HashSet<>(Arrays.asList(arr1)); // Store elements of arr1 in a HashSet.
        for (String word : arr2) {
            if (!set1.contains(word)) { // If any element from arr2 is not in set1, it's not a subset.
                return false;
            }
        }
        return true; // If the loop completes without finding any missing element, arr2 is a subset of arr1.
    }

    // 5. Find the first non-repeating element in an array using LinkedHashMap
    //    -  Time Complexity: O(n), where n is the length of the array.
    //    -  Space Complexity: O(k), where k is the number of unique elements in the array.  In the worst case, k == n.
    public static void findFirstNonRepeating(String[] arr) {
        Map<String, Integer> countMap = new LinkedHashMap<>(); // Use LinkedHashMap to maintain insertion order.

        // Count the frequency of each element in the array.
        for (String word : arr) {
            countMap.put(word, countMap.getOrDefault(word, 0) + 1);
        }

        // Iterate through the map to find the first element with a frequency of 1.
        for (Map.Entry<String, Integer> entry : countMap.entrySet()) {
            if (entry.getValue() == 1) {
                System.out.println(entry.getKey()); // Print the first non-repeating element.
                return; // Exit the method after finding the first non-repeating element.
            }
        }
        System.out.println("No non-repeating element found."); // If no non-repeating element is found.
    }
}

